loader/dd_library_loader.c

```diff
@@ -0,0 +1,355 @@
+
+static bool debug_logs = false;
+static bool force_load = false;
+
+static int php_api_no = 0;
+
+static void ddloader_error_handler(int error_num, zend_string *error_filename, const uint32_t error_lineno, zend_string *message) {}
+
+#define INFO "info"
+#define WARN "warn"
+#define ERROR "error"
+#define LOG(level, format, ...) ddloader_logf(level, format, ##__VA_ARGS__);
+
+static inline void ddloader_logf(const char *level, const char *format, ...) {
+    if (!debug_logs) {
+        return;
+    }
+    char buf[256];
+
+    va_list va;
+    va_start(va, format);
+    vsnprintf(buf, sizeof(buf), format, va);
+    php_printf("[dd_library_loader][%s] %s\n", level, buf);
+    va_end(va);
+}
+
+static char *ddloader_find_ext_path(const char *ext_dir, const char *ext_name, int module_api, bool is_zts, bool is_debug) {
+    char *pkg_path = getenv("DD_LOADER_PACKAGE_PATH");
+    if (!pkg_path) {
+        pkg_path = "/home/circleci/app/dd-library-php";  // FIXME
+    }
+
+    char *full_path;
+    asprintf(&full_path, "%s/%s/ext/%d/%s%s%s.so", pkg_path, ext_dir, module_api, ext_name, is_zts ? "-zts" : "", is_debug ? "-debug" : "");
+
+    if (access(full_path, F_OK)) {
+        free(full_path);
+
+        return NULL;
+    }
+
+    return full_path;
+}
+
+/**
+ * Try to load a symbol from a library handle.
+ * As some OS prepend _ to symbol names, we try to load with and without it.
+ */
+static void *ddloader_dl_fetch_symbol(void *handle, const char *symbol_name_with_underscoe) {
+    void *symbol = DL_FETCH_SYMBOL(handle, symbol_name_with_underscoe + 1);
+    if (!symbol) {
+        symbol = DL_FETCH_SYMBOL(handle, symbol_name_with_underscoe);
+    }
+
+    return symbol;
+}
+
+static bool ddloader_check_deps(const zend_module_dep *deps) {
+    if (!deps) {
+        return true;
+    }
+
+    size_t name_len;
+    zend_string *lcname;
+    char i = 0;
+    while (deps[i].name) {
+        if (deps[i].type == MODULE_DEP_REQUIRED) {
+            zend_module_entry *req_mod;
+
+            name_len = strlen(deps[i].name);
+            lcname = zend_string_alloc(name_len, 0);
+            zend_str_tolower_copy(ZSTR_VAL(lcname), deps[i].name, name_len);
+
+            if ((req_mod = zend_hash_find_ptr(&module_registry, lcname)) == NULL || !req_mod->module_started) {
+                zend_string_efree(lcname);
+                return false;
+            }
+            zend_string_efree(lcname);
+        }
+        ++i;
+    }
+
+    return true;
+}
+
+static void ddloader_unregister_module(void) {
+    zend_module_entry *ddtrace_injected = zend_hash_str_find_ptr(&module_registry, ZEND_STRL("ddtrace_injected"));
+    if (!ddtrace_injected) {
+        return;
+    }
+
+    // Set the MSHUTDOWN function to NULL to avoid it being called by zend_hash_str_del
+    ddtrace_injected->module_shutdown_func = NULL;
+    zend_hash_str_del(&module_registry, ZEND_STRL("ddtrace_injected"));
+}
+
+static PHP_MINIT_FUNCTION(ddtrace_injected) {
```

Comment:
How hard would it be, to not hardcode `"ddtrace"`, but in a single place?
I'd love to be able to later on easily just add `ddloader_load_extension("profiler", /* some more arg which is a hint for ddloader_find_ext_path */)` or similar, and it'll just load the proper extension.

Reply:
Done using a global array
ie.
```
static injected_ext injected_ext_config[] = {
    DECLARE_INJECTED_EXT("ddtrace", "trace", ((zend_module_dep[]){ZEND_MOD_OPTIONAL("json") ZEND_MOD_OPTIONAL("standard") ZEND_MOD_OPTIONAL("ddtrace") ZEND_MOD_END})),
    DECLARE_INJECTED_EXT("datadog-profiling", "profiling", ((zend_module_dep[]){ZEND_MOD_END})),
    DECLARE_INJECTED_EXT("ddappsec", "appsec", ((zend_module_dep[]){ZEND_MOD_END})),
};
```

The complexity is that  we must backup some data of the module when we load it, and restore them at MINIT.
And to avoid creating an MINIT function for each extension we want to load, I use the "module_number" to identify which extension is starting.

Reply:
That's perfect :-)